compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,602 @@
+              (use, defCastBundle));
+          continue;
+        }
+      }
+
+      //denormalize if the def is safe to move and there is no unsafe function
+      //between use and def
+      if(isExprSafeForReorder(def)) {
+        if(!possibleDepInBetween(defPar, usePar)) {
+          denormalize(def, use, castTo);
+        }
+      }
+    }
+  } // end loop for symbol
+
+  if(!cachedGlobalManip) {
+    globalManipFuncCache[fn] = false;
+  }
+
+  //handle deferred actuals
+  for(std::set<CallExpr*>::iterator ceIt = deferredFns.begin() ;
+      ceIt != deferredFns.end() ; ceIt++) {
+    denormalizeActuals(*ceIt, defMap, useMap, actualUseDefMap);
+  }
+
+  freeDefUseMaps(defMap, useMap);
+}
+
+void denormalizeActuals(CallExpr* ce,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap,
+    ActualUseDefCastMap actualUseDefMap) {
+
+  INT_ASSERT(!ce->isPrimitive());
+  for_alist_backward(actual, ce->argList) {
+    if(SymExpr* argSym = toSymExpr(actual)) { //else it's already denormd
+      if(! (argSym->var->isConstant() || argSym->var->isParameter())) {
+        if(actualUseDefMap.count(argSym) > 0) {
+          std::pair<Expr*,Type*> tmpTuple = actualUseDefMap[argSym];
+          SymExpr* use = argSym;
+          Expr* def = tmpTuple.first;
+          Type* castTo = tmpTuple.second;
+          Expr* usePar = use->parentExpr;
+          Expr* defPar = def->parentExpr;
+          if(CallExpr* ceTmp = toCallExpr(defPar)) {
+            if(!isRecord(ceTmp->get(1)->typeInfo())) { //to preserve pass-by-value
+              if(isExprSafeForReorder(def)) {
+                if(!possibleDepInBetween(defPar, usePar)) {
+                  // In C actual evaluation order is not standard, therefore any
+                  // defPar that us unsafe for reordering cannot be moved to
+                  // the args
+                  denormalize(def, use, castTo);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+bool isDenormalizable(Symbol* sym,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap, SymExpr** useOut, Expr** defOut,
+    Type** castTo) {
+
+  if(sym && !(toFnSymbol(sym) || toArgSymbol(sym) || toTypeSymbol(sym))) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar = NULL;
+
+    Vec<SymExpr*>* defs = defMap.get(sym);
+    Vec<SymExpr*>* uses = useMap.get(sym);
+
+    if(defs && defs->n == 1 && uses && uses->n == 1) { // check def-use counts
+      SymExpr* se = defs->first();
+      defPar = se->parentExpr;
+
+      //defPar has to be a move without any coercion
+      CallExpr* ce = toCallExpr(defPar);
+      if(ce) {
+        if(ce->isPrimitive(PRIM_MOVE)) {
+          Type* lhsType = ce->get(1)->typeInfo();
+          Type* rhsType = ce->get(2)->typeInfo();
+          if(lhsType == rhsType) {
+            if(! canPrimMoveCreateCommunication(ce)) {
```

Comment:
Why does denormalizing care if the def can create communication? Is it just that more work would be required in codegen, or is there something fundamental here?